native/extract_subset_extension.cpp

```diff
@@ -0,0 +1,96 @@
+#include "extract_subset_extension.h"
+
+#include <sys/stat.h>
+#include <assert.h>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+namespace py = pybind11;
+
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "constdb.h"
+#include "reader.h"
+#include "writer.h"
+
+class Reader2Writer {
+  public:
+    Reader2Writer(const char * src_path,
+		  float subset_ratio) : reader(src_path, true), iterator(reader.iter()) {
+      size_t total_patients = reader.get_patient_ids().size();
+      size_t target_num_patients = static_cast<size_t>((float) total_patients * subset_ratio);
+      std::cout << "Selecting " << target_num_patients << " out of " << total_patients << " for subset" << std::endl;
+      size_t num_processed = 0;
+      for (auto patient_id : reader.get_patient_ids()) {
+	if (num_processed >= target_num_patients) {
```

Comment:
This should randomly sample instead of taking the first num_patients. You would also need a seed parameter.